addons/survey/controllers/main.py

```diff
@@ -68,18 +68,22 @@ def _check_validity(self, survey_token, answer_token, ensure_token=True, check_p
         if answer_token and not answer_sudo:
             return 'token_wrong'
 
-        if not answer_sudo and ensure_token:
-            return 'token_required'
-        if not answer_sudo and survey_sudo.access_mode == 'token':
-            return 'token_required'
+        if ensure_token != 'allow_no_answer_token_mode':
+            if not answer_sudo and ensure_token:
+                return 'token_required'
+            if not answer_sudo and survey_sudo.access_mode == 'token':
+                return 'token_required'
```

Comment:
There is something I don't get. Is there a case where passing `ensure_token=False` should return `token_required` because of `access_mode`when not coming from print?

If there isn't, we can drop the check (73-74L) altogether.
If there is one, I would rename `'allow_no_answer_token_mode'` to `'survey_only'`, and adapt the checks like follows:

```suggestion     
        if (not answer_sudo
            and (ensure_token is True
                 or (ensure_token != 'survey_only' and survey_sudo.access_mode == 'token'))
        ):
            return 'token_required'
```